rpcs3/Emu/RSX/color_utils.h

```diff
@@ -0,0 +1,245 @@
+					break;
+				case CELL_GCM_TEXTURE_REMAP_ZERO:
+					remapped[channel] = select_zero;
+					break;
+				case CELL_GCM_TEXTURE_REMAP_ONE:
+					remapped[channel] = select_one;
+					break;
+				}
+			}
+			return remapped;
+		}
+
+		template <typename T>
+			requires std::is_integral_v<T> || std::is_floating_point_v<T>
+		std::array<T, 4> remap(const std::array<T, 4>& components) const
+		{
+			return remap(components, static_cast<T>(0), static_cast<T>(1));
+		}
+
+		template <typename T>
+		color4_base<T> remap(const color4_base<T>& components)
+		{
+			const std::array<T, 4> values = { components.a, components.r, components.g, components.b };
+			const auto shuffled = remap(values, T{ 0 }, T{ 1 });
+			return color4_base<T>(shuffled[1], shuffled[2], shuffled[3], shuffled[0]);
+		}
+
+		template <typename T>
+			requires std::is_integral_v<T> || std::is_enum_v<T>
+		texture_channel_remap_t with_encoding(T encoding) const
+		{
+			texture_channel_remap_t result = *this;
+			result.encoded = encoding;
+			return result;
+		}
+	};
+
+	static const texture_channel_remap_t default_remap_vector =
+	{
+		.encoded = RSX_TEXTURE_REMAP_IDENTITY,
+		.control_map = { CELL_GCM_TEXTURE_REMAP_REMAP, CELL_GCM_TEXTURE_REMAP_REMAP, CELL_GCM_TEXTURE_REMAP_REMAP, CELL_GCM_TEXTURE_REMAP_REMAP },
+		.channel_map = { CELL_GCM_TEXTURE_REMAP_FROM_A, CELL_GCM_TEXTURE_REMAP_FROM_R, CELL_GCM_TEXTURE_REMAP_FROM_G, CELL_GCM_TEXTURE_REMAP_FROM_B }
+	};
+
+	static inline texture_channel_remap_t decode_remap_encoding(u32 remap_ctl)
+	{
+		// Remapping tables; format is A-R-G-B
+		// Remap input table. Contains channel index to read color from
+		texture_channel_remap_t result =
+		{
+			.encoded = remap_ctl
+		};
+
+		result.channel_map =
+		{
+			static_cast<u8>(remap_ctl & 0x3),
+			static_cast<u8>((remap_ctl >> 2) & 0x3),
+			static_cast<u8>((remap_ctl >> 4) & 0x3),
+			static_cast<u8>((remap_ctl >> 6) & 0x3),
+		};
+
+		// Remap control table. Controls whether the remap value is used, or force either 0 or 1
+		result.control_map =
+		{
+			static_cast<u8>((remap_ctl >> 8) & 0x3),
+			static_cast<u8>((remap_ctl >> 10) & 0x3),
+			static_cast<u8>((remap_ctl >> 12) & 0x3),
+			static_cast<u8>((remap_ctl >> 14) & 0x3),
+		};
+
+		return result;
+	}
+
+	// Convert color write mask for G8B8 to R8G8
+	static inline u32 get_g8b8_r8g8_clearmask(u32 mask)
+	{
+		u32 result = 0;
+		if (mask & RSX_GCM_CLEAR_GREEN_BIT) result |= RSX_GCM_CLEAR_GREEN_BIT;
+		if (mask & RSX_GCM_CLEAR_BLUE_BIT) result |= RSX_GCM_CLEAR_RED_BIT;
+
+		return result;
+	}
+
+	static inline void get_g8b8_r8g8_colormask(bool& red, bool&/*green*/, bool& blue, bool& alpha)
+	{
+		red = blue;
+		blue = false;
+		alpha = false;
+	}
+
+	static inline void get_g8b8_clear_color(u8& red, u8& /*green*/, u8& blue, u8& /*alpha*/)
+	{
+		red = blue;
+	}
+
+	static inline u32 get_abgr8_clearmask(u32 mask)
+	{
+		u32 result = 0;
+		if (mask & RSX_GCM_CLEAR_RED_BIT) result |= RSX_GCM_CLEAR_BLUE_BIT;

```

Comment:
is this correct? also see BLUE and RED

Reply:
Yea, it's correct, just weird. There is no A-B-G-R format so we need to swap R and B to make  it A-R-G-B a.k.a BGRA8

Reply:
Most of this file is battle-tested, I just removed color-related stuff from the other overcrowded utils header